OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs

```diff
@@ -11,55 +11,146 @@
 using System.Collections.Generic;
 using System.Drawing;
 using System.Linq;
+using OpenRA.Graphics;
 using OpenRA.Mods.Common.Traits;
-using OpenRA.Primitives;
 using OpenRA.Widgets;
 
 namespace OpenRA.Mods.Common.Widgets
 {
 	public class SupportPowerTimerWidget : Widget
 	{
-		public readonly string Font = "Bold";
-		public readonly string Format = "{0}: {1}";
-		public readonly TimerOrder Order = TimerOrder.Descending;
+		[Translate] public readonly string NoTeamText = "No Team";
+		[Translate] public readonly string TeamText = "Team {0}";
 
-		readonly IEnumerable<SupportPowerInstance> powers;
-		Pair<string, Color>[] texts;
+		readonly Dictionary<Player, SupportPowerManager> spManagers;
+		readonly SortedDictionary<int, List<Player>> playersByTeam;
+
+		readonly string fontName = "Bold";
+		readonly SpriteFont font;
+
+		readonly float yIncrement;
+		readonly float padding;
+
+		struct Text
+		{
+			public Color Color;
+			public string Team;		// Used only for team label
+			public string Player;
+			public string Power;
+			public string RemainingTime;
+		}
+
+		List<Text> display;
+		float maxPlayerWidth;
+		float maxPowerWidth;
+
+		World world;
+		int lastTick;
 
 		[ObjectCreator.UseCtor]
 		public SupportPowerTimerWidget(World world)
 		{
-			powers = world.ActorsWithTrait<SupportPowerManager>()
-				.Where(p => !p.Actor.IsDead && !p.Actor.Owner.NonCombatant)
-				.SelectMany(s => s.Trait.Powers.Values)
-				.Where(p => p.Instances.Any() && p.Info.DisplayTimer && !p.Disabled);
+			spManagers = new Dictionary<Player, SupportPowerManager>();
+			foreach (var tp in world.ActorsWithTrait<SupportPowerManager>().Where(tp => !tp.Actor.Owner.NonCombatant))
+				spManagers[tp.Actor.Owner] = tp.Trait;
```

Comment:
That won't work, because the instances will change as the game progresses.  What you *can* do is query all the SupportPowerInfos and then store a dictionary of Info to width.